### PR TITLE
optenv32 Survey: Towards the Steam Native Runtime (Wave 1: Reworked Toolchain & Base Libraries)

### PR DESCRIPTION
**Pull request by Cyanoxygen**
Topic Description
-----------------

This PR overhauls the entire 32-bit (i686) subsystem of AOSC OS (optenv32), and make it complete enough to support the Steam Native Runtime.

Package(s) Affected
-------------------

TBD

Security Update?
----------------

No

Prerequisites
----------

- [x] AOSC-Dev/autobuild4#10

Build Order
-----------

### Base packages

Base packages should be built in order, manual intervention REQUIRED:

#### Bootstrapping the toolchain

```
ciel build autobuild4 binutils+32:+stage2 gcc+32:+stage2 linux+api+32:+stage2 glibc+32:+stage2 libxcrypt
```

And then, refresh the local package repo, and start the container, install the packages, then build stage 2 gcc+32 again. This time the stage 2 GCC will be complete enough to build the stage 3 GCC, since it then ships libstdc++.

```
ciel repo init <instance>
ciel sh -i <instance>
apt update
apt install autobuild4 gcc+32 glibc+32
acbs-build gcc+32:+stage2
```

#### Base libraries: Wave 1

```
ciel build zlib+32 gmp+32 mpfr+32 mpc+32 binutils+32 gcc+32 glibc+32 libxcrypt+32 gcc+32 ncurses+32 readline+32 expat+32 icu+32 xz+32 bzip2+32 db+32
```

Test Build(s) Done
------------------

**Primary Architectures**

- [x] AMD64 `amd64`
- [x] 32-bit Optional Environment `optenv32`

### CODE REVIEW

**Comment by Cyanoxygen** [comment]
After building glibc+32 with `-mstackrealign`, Steam is able to start games, the games run perfectly as expected.

Proton games are not tested though.

**Comment by jiegec** [comment]
Can't compile i686 binary manually:

```shell
$ /opt/32/bin/g++ test.cpp -o test
/usr/bin/ld: /opt/32/lib/gcc/i686-aosc-linux-gnu/13.2.0/liblto_plugin.so: error loading plugin: /opt/32/lib/gcc/i686-aosc-linux-gnu/13.2.0/liblto_plugin.so: wrong ELF class: ELFCLASS32
collect2: error: ld returned 1 exit status
```

**Comment by Cyanoxygen** [comment]
> Can't compile i686 binary manually:
> 
> ```shell
> $ /opt/32/bin/g++ test.cpp -o test
> /usr/bin/ld: /opt/32/lib/gcc/i686-aosc-linux-gnu/13.2.0/liblto_plugin.so: error loading plugin: /opt/32/lib/gcc/i686-aosc-linux-gnu/13.2.0/liblto_plugin.so: wrong ELF class: ELFCLASS32
> collect2: error: ld returned 1 exit status
> ```

hmmm, must find a way to tell gcc to find the correct linker. For now please `export PATH=/opt/32/bin:$PATH`.

File: runtime-optenv32/binutils+32/autobuild/beyond [hunk 1]
```diff
@@ -0,0 +1,20 @@
+abinfo "Installing libiberty header ..."
+cp -v include/libiberty.h "$PKGDIR"/opt/32/include
```
Comment (MingcongBai) [inline runtime-optenv32/binutils+32/autobuild/beyond]: "$SRCDIR"/include/libiberty.h

File: runtime-optenv32/binutils+32/autobuild/beyond.stage2 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+abinfo "Installing libiberty.h ..."
+install -Dvm644 "$SRCDIR"/include/libiberty.h "$PKGDIR"/opt/32/include/libiberty.h
+
+abinfo "Dropping unneeded files ..."
+rm -r "$PKGDIR"/opt/32/share
```
Comment (MingcongBai) [inline runtime-optenv32/binutils+32/autobuild/beyond.stage2]: Verbosity here.

File: runtime-optenv32/binutils+32/autobuild/prepare [hunk 3]
```diff
@@ -0,0 +1,4 @@
+sed -i "/ac_cpp=/s/\$CPPFLAGS/\$CPPFLAGS -O2/" "$SRCDIR"/libiberty/configure
```
Comment (MingcongBai) [inline runtime-optenv32/binutils+32/autobuild/prepare:4]: Lint this script.

File: runtime-optenv32/devel-base+32/autobuild/build [hunk 4]
```diff
@@ -0,0 +1 @@
+mkdir "$PKGDIR"
```
Comment (MingcongBai) [inline runtime-optenv32/devel-base+32/autobuild/build]: Verbosity here.

Comment (MingcongBai) [reply runtime-optenv32/devel-base+32/autobuild/build]: Also, why not `ABTYPE=dummy`?

File: runtime-optenv32/gcc+32/autobuild/prepare.stage2 [hunk 5]
```diff
@@ -0,0 +1,11 @@
+abinfo "Extracting gmp..."
+tar xf "$SRCDIR"/../gmp-${__GMP_VER}.tar.xz
```
Comment (MingcongBai) [inline runtime-optenv32/gcc+32/autobuild/prepare.stage2]: Verbosity for all `tar` commands.

File: runtime-optenv32/glibc+32/autobuild/beyond [hunk 6]
```diff
@@ -1,6 +1,3 @@
-mkdir -p "$PKGDIR"/opt/32/lib
-
-rm -rf "$PKGDIR"/32
-rm -rf "$PKGDIR"/opt/32/lib/locale
-
-rm -rf "$PKGDIR"/opt/32/{var,etc,libexec,share}
+abinfo "Creating symbol link dynamic loader in the main system ..."
+mkdir -pv "$PKGDIR"/usr/lib
+ln -sv /opt/32/lib/ld-linux.so.2 "$PKGDIR"/usr/lib/ld-linux.so.2
```
Comment (MingcongBai) [inline runtime-optenv32/glibc+32/autobuild/beyond]: Use relative path.

File: runtime-optenv32/gcc+32/autobuild/build.stage2 [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+fi
+
+# Linux From Scratch: Since glibc might not yet available, we need a proper
+# limits.h header to support further builds.
+abinfo "Fixing limits.h header ..."
+cat "$SRCDIR"/gcc/limitx.h "$SRCDIR"/gcc/glimits.h "$SRCDIR"/gcc/limity.h > \
+	`dirname $("$PKGDIR"/opt/32/bin/i686-aosc-linux-gnu-gcc -print-libgcc-file-name)`/include/limits.h
+
+abinfo "Dropping unneeded files ..."
+if [ -d "$PKGDIR"/opt/32/share ] ; then
```
Comment (MingcongBai) [inline runtime-optenv32/gcc+32/autobuild/build.stage2]: Do you not know if these exist?

File: runtime-optenv32/glibc+32/autobuild/build.stage2 [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+abinfo "Building stage 2 glibc ..."
+make
+
+abinfo "Installing stage 2 glibc ..."
+make install_root="$PKGDIR" DESTDIR="$PKGDIR" install 
+
+cd "$SRCDIR"
+
+mkdir -pv "$PKGDIR"/usr/lib
+ln -sv /opt/32/lib/ld-linux.so.2 "$PKGDIR"/usr/lib/ld-linux.so.2
```
Comment (MingcongBai) [inline runtime-optenv32/glibc+32/autobuild/build.stage2]: Relative path here.

File: runtime-optenv32/glibc+32/autobuild/build.stage2 [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                 --cache-file=config.cache \
+                 libc_cv_slibdir=/opt/32/lib
+
+abinfo "Building stage 2 glibc ..."
+make
+
+abinfo "Installing stage 2 glibc ..."
+make install_root="$PKGDIR" DESTDIR="$PKGDIR" install 
+
+cd "$SRCDIR"
```
Comment (MingcongBai) [inline runtime-optenv32/glibc+32/autobuild/build.stage2]: This is not necessary.

File: runtime-optenv32/glibc+32/autobuild/prepare.stage2 [hunk 10]
```diff
@@ -0,0 +1,6 @@
+export PATH=/opt/32/bin:$PATH
+export CC=/opt/32/bin/i686-aosc-linux-gnu-gcc
+export CXX=/opt/32/bin/i686-aosc-linux-gnu-g++
+export LD=/opt/32/bin/i686-aosc-linux-gnu-ld.bfd
+unset CPPFLAGS CFLAGS CXXFLAGS LDFLAGS
+
```
Comment (MingcongBai) [inline runtime-optenv32/glibc+32/autobuild/prepare.stage2]: Trailing blank line.

File: runtime-optenv32/libxcrypt+32/autobuild/build [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+build_autotools_configure
+
+abinfo "Building libxcrypt (new API) ..."
+build_autotools_build
+
+abinfo "Installing libxcrypt (new API) ..."
+build_autotools_install
+
+abinfo "Resetting source tree ..."
+rm -r "$BLDDIR"
```
Comment (MingcongBai) [inline runtime-optenv32/libxcrypt+32/autobuild/build]: Verbosity here.

File: runtime-optenv32/linux+api+32/autobuild/defines.stage2 [hunk 12]
```diff
@@ -0,0 +1,5 @@
+PKGNAME=linux+api+32
+PKGSEC=devel
+PKGDES="Linux API Headers for glibc (optenv32)"
+NOLTO=1
```
Comment (MingcongBai) [inline runtime-optenv32/linux+api+32/autobuild/defines.stage2]: What...?

File: runtime-optenv32/linux+api+32/spec [hunk 13]
```diff
@@ -1,5 +1,8 @@
-VER=4.5.2
-SRCS="tbl::https://cdn.kernel.org/pub/linux/kernel/v${VER:0:1}.x/linux-$VER.tar.xz"
-CHKSUMS="sha256::e32339db8365f0f10b873928095da47dd3e8a40f8ea5edd5501320d4ff30209f"
-REL=2
-CHKUPDATE="git::url=https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git;pattern=v5.4.*"
+VER=6.7.2
```
Comment (MingcongBai) [inline runtime-optenv32/linux+api+32/spec]: Sync with main tree.

File: runtime-optenv32/ncurses+32/autobuild/beyond [hunk 14]
```diff
@@ -0,0 +1,11 @@
+abinfo "Appending wide-character linkage to pkg-config files..."
+for lib in ncurses ncurses++ form panel menu; do
+	echo "INPUT(-l${lib}w)" > "$PKGDIR"/opt/32/lib/lib${lib}.so
+	ln -s ${lib}w.pc "$PKGDIR"/opt/32/lib/pkgconfig/${lib}.pc
```
Comment (MingcongBai) [inline runtime-optenv32/ncurses+32/autobuild/beyond]: Verbosity here.

File: runtime-optenv32/readline+32/spec [hunk 15]
```diff
@@ -1,5 +1,4 @@
-VER=6.3.008
-SRCS="tbl::https://ftp.gnu.org/gnu/readline/readline-${VER:0:3}.tar.gz"
-CHKSUMS="sha256::56ba6071b9462f980c5a72ab0023893b65ba6debb4eeb475d7a563dc65cafd43"
-REL=5
+VER=8.2.1
+SRCS="https://ftp.gnu.org/gnu/readline/readline-${VER:0:3}.tar.gz"
+CHKSUMS="sha256::3feb7171f16a84ee82ca18a36d7b9be109a52c04f492a053331d7d1095007c35"
```
Comment (MingcongBai) [inline runtime-optenv32/readline+32/spec:3]: You may want to create symlinks for 6 and 7.

File: runtime-optenv32/zlib+32/autobuild/build [hunk 16]
```diff
@@ -1,8 +1,9 @@
-export PATH=/opt/32/bin:$PATH
-CC=i686-pc-linux-gnu-gcc CXX=i686-pc-linux-gnu-g++ \
-AS=i686-pc-linux-gnu-as LD=i686-pc-linux-gnu-ld AR=i686-pc-linux-gnu-ar \
-RANLIB=i686-pc-linux-gnu-ranlib ./configure --prefix=/opt/32
+CC=i686-aosc-linux-gnu-gcc \
```
Comment (MingcongBai) [inline runtime-optenv32/zlib+32/autobuild/build:2]: Lint this script.

File: runtime-optenv32/zlib+32/spec [hunk 17]
```diff
@@ -1,5 +1,4 @@
-VER=1.2.11
-REL=2
-SRCS="tbl::https://www.zlib.net/zlib-$VER.tar.xz"
-CHKSUMS="sha256::4ff941449631ace0d4d203e3483be9dbc9da454084111f97ea0a2114e19bf066"
+VER=1.3.1
+SRCS="http://zlib.net/zlib-$VER.tar.xz"
```
Comment (MingcongBai) [inline runtime-optenv32/zlib+32/spec]: No HTTPS?

**Review by MingcongBai — CHANGES_REQUESTED** [review]

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**